README.md
#### NAME / НАЗВАНИЕ

phpapi: The PHP Documentation Creator.
This is the fork of the Peej's Quick & Dirty PHPDoc Clone.

#### VERSION / ВЕРСИЯ

5.0

#### DESCRIPTION / ОПИСАНИЕ

phpapi is a Javadoc style comment parser for PHP, written with an emphasis on speed and simplicity.
phpapi - это парсер комментариев в стиле Javadoc для PHP, написан с акцентом на скорость и простоту.

#### REQUIREMENTS / ТРЕБОВАНИЯ

PHP 5.4+ is required.

#### ABOUT / О ПРОГРАММЕ

phpapi uses the PHP tokenizer extension, parses the tokens and translates them into API documentation.
phpapi использует PHP-расширение tokenizer, анализирует токены PHP и транслирует их в документацию по API.

#### FEATURES / ВОЗМОЖНОСТИ

- Fast speed.
- Uses PHP tokenizer to take advantage of PHPs internal parsing functionality.
- Parsing of any valid PHP file, with multiple classes and functions in the same file.
- Simple to install and use, instant results.
- Documents global constants.
- Original template system for output.


- Высокая скорость.
- Использует PHP tokenizer для облегчения внутреннего разбора функциональности.
- Парсит любой корректный PHP-файл, с несколькими классами и функциями в одном файле.
- Простой в установке и использовании, мгновенный результат.
- Документирует глобальные константы.
- Оригинальная система шаблонов для вывода.

#### FILES / ФАЙЛЫ

    phpapi.php            - main executable
    phpapi.ini            - Default config file
    README.md             - This file
    classes/              - Classes used by phpapi
    doclets/html/frames/  - HTML doclet for output with frames
    doclets/html/plain/   - HTML doclet for output without frames
    docs/                 - some docs and licenses
    formatters/           - Formatters
    locales/              - Translation files
    resources/            - Illustrations, javascript and so on
    templates/frames/     - Templates for output with frames
    templates/plain/      - Templates for output without frames

#### INSTALLATION / УСТАНОВКА

Unzip the archive somewhere, edit the config file and then run phpapi.php.
Распакуйте архив, отредактируйте конфигурационный файл и запустите phpapi.php.

#### USAGE / ИСПОЛЬЗОВАНИЕ

    ./phpapi.php (if your config file is phpapi.ini)
    ./phpapi.php config.ini (if your config file is config.ini)
    ./phpapi.php config.ini > result.txt (To save output mesages)

To create a config file for your project, copy, rename the phpini.ini to project.ini file and edit it to your needs, it's fully commented.
Для созданияя конфигурационного файла вашего проекта, скопируйте, переименуйте phpapi.ini в project.ini и отредактируйте его в соответствие с имеющемися комментариями.

#### CONFIGURATION / КОНФИГУРАЦИЯ

phpapi supports a number of configuration directives:
phpapi поддерживает следующие директивы:

    source           - The directory to look for files in,
    destination      - The directory to place generated documentation in.
    files            - Names or wildcards of files to parse.
    ignore             Names of files or directories to ignore. Wildcards are NOT allowed.
    subDirs = on|off - To look in each sub directory for files.
    generator        - Documentation generator.
    formatter        - Documentation formatter.
    doclet           - The doclet to use for generating output.
    verbose = on|off - Verbose mode outputs additional messages during execution.
    defaultPackage   - It for elements without @package tag.
    overview         - The name of file containing overview text to be placed on the overview page.
    windowtitle      - The title to be placed in the HTML page.
    doctitle         - The title to be placed near the top of the overview summary file.
    header           - The header text to be placed at the top of each output file.

#### DOC COMMENTS

A full description of the format of doc comments can be found on the Sun Javadoc web site [javadoc](http://java.sun.com/j2se/javadoc/) .
Doc comments for phpAPI are look like this:

/** <br />
 \* This is the typical format of a simple documentation comment<br />
 \* in two lines.<br />
 \* <br />
 \* __@param__ string $var Description<br />
 \*/

Class fields may be commented as:

/** Comment.<br />
 \* __@var__ integer<br />
 \*/

or

/** __@var__ integer Comment */

#### TAGS / ТЭГИ

phpapi supports the following tags:
phpapi поддерживает следующие теги:

    @abstract                    - defining a class or method as abstract
    @access type                 - the access type of the field or method
    @author name                 - author name, email, web...
    @deprecated                  - deprecated element
    @final                       - defining a class or method as final
    { @link package.class#member label }      - link to element
    { @linkplain package.class#member label } - plain link to element
    @overview                    - detailed description of a package
    @package name                - package name
    @param type name description - description of method or function parameters
    @private                     - element is private
    @program                     - The title of the application, script or libruary
    @protected                   - element is protected
    @public                      - element is public
    @return type description     - return value
    @see package.class#member    - link to corresponding information
    @since                       - since-text
    @static                      - defining a class, method or member variable as static
    @throws                      - exeptions
    @var type name description   - variable description
    @version                     - version number

#### COPYRIGHT AND LICENSE

This program is a fork of the PHPDoctor: The PHP Documentation Creator v2.0.5

Copyright (C) 2005 Paul James <paul@peej.co.uk>
Copyright (C) 2016 Victor Nabatov <greenray.spb@gmail.com>

This program is free software.
You can redistribute it and/or modify it under the terms of the [Creative Commons Attribution-ShareAlike 4.0 International License](https://creativecommons.org/licenses/by-sa/4.0/legalcode) .
This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Эта программа является свободной.
Вы можете распространять и/или модифицировать ее в соответствии c условиями [Creative Commons Attribution-ShareAlike 4.0 International License](https://creativecommons.org/licenses/by-sa/4.0/legalcode) .
Эта программа распространяется в надежде что она будет полезной, но БЕЗ КАКИХ-ЛИБО ГАРАНТИЙ;
даже без подразумеваемых гарантий КОММЕРЧЕСКОЙ ЦЕННОСТИ или ПРИГОДНОСТИ ДЛЯ КОНКРЕТНОЙ ЦЕЛИ.


docs/ChangeLog
09.06.2016
- полностью переписана система шаблонов.
29.04.2015
- новая система шаблонов.
20.04.2015
- добавлена возможность обработка директив include, include_once, require, require_once (сыровато, но работает);
- теперь программа не привязана к абсолютным путям;
- какая-то херня возникла с обработкой исключительно файла template.php - сбрасывается шаблон на функции __if-else.
  И вообще шаблонизатор сложноват и не гибок. Задумал его переработать.
- почистил код и комментарии к нему;
- версия 4.1
14.04.2015:
- добавлен тег @overview - описание пакета (см. код);
- добавлена обработка и отображение в документации ссылок на элементы;
- добавлен парсер разметки markdown;
- полная переработка - новая версия 4.0
02.04.2015
- переработан doclet htmlNoFrames: уменьшен объем генерируемых файлов на 30%, улучшена навигация, увеличина скорость работы;
- добавлено несколько новых шаблонов;
- новый релиз 3.1
30.03.2015
- новый генератор документации - htmlNoFrames;
- значения констант и полей теперь выводятся полностью;
- новый релиз 3.0
22.03.2015
- завершено создание шаблонов для страниц документации;
- новый релиз 2.0
17.03.2015
- переход к использованию шаблонов;
- Нет достаточниых оснований включать исходные тексты в документацию: отключено;
- соответственно отключен geshi;
- опция use_class_path_as_package не добавляет информативности: отключено;
- описание пакетов в формате html чуть ли не в каждом каталоге проекта представляется излишней нагрузкой на клавиатуру.
  Достаточно разместить необходимую информацию в блоке docComment и она никуда не денется - package_comment_dir : отключено;
- обработка public - private - protected сделана независимой;
- добавлена локализация интерфейса.
07.02.2015
- переход на php 5.6;
- добавлено указание типа аргументов функций;
- добавлена подсветка переменных, типов, аргументов, значений констант и переменных;
- скорректирован дизайн;

21.12.2014 - начало работы.

======================
Greenray
greenray.spb@gmail.com


phpapi.php
#!/usr/bin/php
<?php
/**
 * phpapi: PHP Documentation Creator
 * Copyright (c) 2015-2016 Victor Nabatov <greenray.spb@gmail.com>
 *
 * This program is a fork of the
 * PHPDoctor: PHP Documentation Creator version 2.0.5
 * Copyright (C) 2005 Paul James <paul@peej.co.uk>
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * Creative Commons Attribution-ShareAlike 4.0 International License
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * @program   phpapi: PHP Documentation Creator
 * @version   6.0
 * @author    Victor Nabatov greenray.spb@gmail.com
 * @copyright (c) 2015-2016 Victor Nabatov
 * @license   Creative Commons Attribution-ShareAlike 4.0 International
 * @file      phpapi.php
 * @package   phpapi
 */
#
# Check we are running from the command line
#
if (!isset($argv[0])) {
    die('This program must be run from the command line using the CLI version of PHP.');
}
#
# Check we are using the correct version of PHP
#
elseif (version_compare(phpversion(), '5.4.1', '<') || !extension_loaded('tokenizer')) {
    die('You need PHP version 5.4 or greater with the "tokenizer" extension to run this script, please upgrade.');
}

ini_set('display_errors', 1);
mb_internal_encoding('UTF-8');
setlocale(LC_CTYPE, ['ru_RU.UTF-8', 'en_US.UTF-8', 'en_GB.UTF-8']);
setlocale(LC_ALL,   ['ru_RU.UTF-8', 'en_US.UTF-8', 'en_GB.UTF-8']);

error_reporting(E_ALL & ~E_DEPRECATED);

/** Alias for DIRECTORY_SEPARATOR */
define('DS', DIRECTORY_SEPARATOR);
/** Alias for line feed */
define('LF', PHP_EOL);

/** System root */
define('ROOT', '.'.DS);
/** System */
define('SYSTEM',     ROOT.'system'.DS);
/** System classes */
define('CLASSES',    SYSTEM.'classes'.DS);
/** System locales */
define('LOCALES',    ROOT.'locales'.DS);
/** Images, javascripts and other resources */
define('RESOURCES',  ROOT.'resources'.DS);
/** Taglets */
define('TAGLETS',    SYSTEM.'taglets'.DS);
/** Templates */
define('TEMPLATES',  ROOT.'templates'.DS);
/** Markdown formatter */
define('MARKDOWN',   SYSTEM.'markdown'.DS);
/** Standard error output. */
if (!defined('STDERR')) define('STDERR', fopen("php://stderr", "wb"));

/** Version of the system */
define('VERSION', '6.0');
/** Copyright */
define('COPYRIGHT', '&copy; 2015 - 2016 Greenray');
/** System generator */
define('GENERATOR', 'Generated by <a href="https://github.com/Greenray/phpAPI" target="_blank">phpAPI</a>: PHP Documentation Creator ver. '.VERSION);

require CLASSES.'doc.php';
require CLASSES.'rootDoc.php';
require CLASSES.'elementDoc.php';
require CLASSES.'packageDoc.php';
require CLASSES.'classDoc.php';
require CLASSES.'methodDoc.php';
require CLASSES.'fieldDoc.php';
require CLASSES.'type.php';
require TAGLETS.'tag.php';
require CLASSES.'template.php';
require CLASSES.'phpAPI.php';

/** Website localization. */
global $lang;

/**
 * String localization.
 * Currently, the system supports two languages: English and Russian.
 *
 * @global  array  $lang   Array of language strings
 * @param   string $string String to be translated
 * @return  string         Translated string
 * @package phpapi
 */
function __($string) {
    global $lang;
    return empty($lang[$string]) ? $string : $lang[$string];
}

#
# Get name of config file to use
#

if (!isset($argv[1])) {
    if     (isset($_ENV['phpapi']))                     $argv[1] = $_ENV['phpapi'];
    elseif (is_file(getcwd().DS.'phpapi.ini'))          $argv[1] = getcwd().DS.'phpapi.ini';
    elseif (is_file(dirname(__FILE__).DS.'phpapi.ini')) $argv[1] = dirname(__FILE__).DS.'phpapi.ini';
    else {
        die("Usage: phpapi [config_file]".LF);
    }
}

$phpdoc  = new phpapi($argv[1]);
$rootDoc = $phpdoc->parse();
$phpdoc->execute($rootDoc);
